kernel/libs/aster-bigtcp/src/socket/bound.rs

```diff
@@ -264,50 +320,67 @@ impl Deref for NeedIfacePoll {
     }
 }
 
-impl<E: Ext> BoundTcpSocket<E> {
+impl<E: Ext> TcpConnection<E> {
     /// Connects to a remote endpoint.
     ///
     /// Polling the iface is _always_ required after this method succeeds.
-    pub fn connect(
-        &self,
+    pub fn new_connect(
+        bound: BoundPort<E>,
         remote_endpoint: IpEndpoint,
-    ) -> Result<(), smoltcp::socket::tcp::ConnectError> {
-        let common = self.iface().common();
-        let mut iface = common.interface();
+        option: &RawTcpOption,
+        observer: E::TcpEventObserver,
+    ) -> Result<Self, (BoundPort<E>, smoltcp::socket::tcp::ConnectError)> {
+        let socket = {
+            let mut socket = new_tcp_socket();
 
-        let mut socket = self.0.socket.lock();
+            option.apply(&mut socket);
 
-        socket.connect(iface.context(), remote_endpoint, self.0.port)?;
+            let common = bound.iface().common();
+            let mut iface = common.interface();
 
-        socket.has_connected = false;
-        self.0.update_next_poll_at_ms(PollAt::Now);
+            if let Err(err) = socket.connect(iface.context(), remote_endpoint, bound.port()) {
+                drop(iface);
+                return Err((bound, err));
+            }
+
+            socket
+        };
 
-        Ok(())
+        let inner = TcpConnectionInner::new(socket, None);
+
+        let this = Self::new(bound, inner);
+        this.0.update_next_poll_at_ms(PollAt::Now);
+        this.init_observer(observer);
+        this.iface()
+            .common()
+            .register_tcp_connection(this.inner().clone());
+
+        Ok(this)
     }
 
     /// Returns the state of the connecting procedure.
     pub fn connect_state(&self) -> ConnectState {
-        let socket = self.0.socket.lock();
+        let socket = self.0.inner.lock();
 
         if socket.state() == State::SynSent || socket.state() == State::SynReceived {
             ConnectState::Connecting
         } else if socket.has_connected {
             ConnectState::Connected
+        } else if KeyableArc::strong_count(self.0.as_ref()) > 1 {
+            ConnectState::Connecting
```

Comment:
This is a rather magic condition check. How about adding some comments to explain why it holds? For example, who will holds `self.0` when connecting?

Reply:
I added a few comments here.